lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemStats.java

```diff
@@ -0,0 +1,264 @@
+    @Nested
+    public AwsSdkV2ApiCallStats getAbortMultipartUpload()
+    {
+        return abortMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats uploadPart()
+    {
+        return uploadPart;
+    }
+
+    public MetricPublisher newMetricPublisher()
+    {
+        return new JmxMetricPublisher(this);
+    }
+
+    public static final class JmxMetricPublisher
+            implements MetricPublisher
+    {
+        private static final Set<SdkMetric<?>> ALLOWED_METRICS = Set.of(API_CALL_SUCCESSFUL, RETRY_COUNT, API_CALL_DURATION, ERROR_TYPE);
+
+        private static final Logger log = Logger.get(JmxMetricPublisher.class);
+
+        private final S3FileSystemStats stats;
+
+        public JmxMetricPublisher(S3FileSystemStats stats)
+        {
+            this.stats = requireNonNull(stats, "stats is null");
+        }
+
+        @Override
+        public void publish(MetricCollection metricCollection)
+        {
+            try {
+                Optional<String> serviceId = metricCollection.metricValues(SERVICE_ID).stream().filter(Objects::nonNull).findFirst();
+                Optional<String> operationName = metricCollection.metricValues(OPERATION_NAME).stream().filter(Objects::nonNull).findFirst();
+                if (serviceId.isEmpty() || operationName.isEmpty()) {
+                    log.warn("ServiceId or OperationName is empty for AWS MetricCollection: %s", metricCollection);
+                    return;
+                }
+
+                if (!serviceId.get().equals("S3")) {
+                    return;
+                }
+
+                AwsSdkV2ApiCallStats apiCallStats = getApiCallStats(operationName.get());
+                publishMetrics(metricCollection, apiCallStats);
+            }
+            catch (Exception e) {
+                log.warn(e, "Publishing AWS metrics failed");
+            }
+        }
+
+        private void publishMetrics(MetricCollection metricCollection, AwsSdkV2ApiCallStats apiCallStats)
+        {
+            metricCollection.stream()
+                    .filter(metricRecord -> metricRecord.metric() != null && metricRecord.value() != null)
+                    .filter(metricRecord -> ALLOWED_METRICS.contains(metricRecord.metric()))
+                    .forEach(metricRecord -> {
+                        if (metricRecord.metric().equals(API_CALL_SUCCESSFUL)) {
+                            Boolean value = (Boolean) metricRecord.value();
+
+                            stats.total.updateCalls();
+                            apiCallStats.updateCalls();
+
+                            if (value.equals(Boolean.FALSE)) {
+                                stats.total.updateFailures();
+                                apiCallStats.updateFailures();
+                            }
+                        }
+                        else if (metricRecord.metric().equals(RETRY_COUNT)) {
+                            Integer value = (Integer) metricRecord.value();
```

Comment:
Similar to the above, this value is going to be unboxed on both calls to `updateRetries` below- so if this value could be `null` than we should check for that, and in either case we should unbox to `int` once and pass the primitive into the `updateRetries` calls.

Reply:
Ack. Updated in the revision.